### PR TITLE
Initial implementation of transition_hyperparameters

### PR DESCRIPTION
**Pull request by ThomasColthurst**
Only on the Distribution side (I.e., not hooked up to hirm.cc yet) and only for BetaBernoulli for now.

### CODE REVIEW

File: cxx/distributions/beta_bernoulli.hh [hunk 1]
```diff
@@ -8,42 +8,68 @@
...
+        assert(0 <= N);
+    }
+    double logp(const double& x) const {
+        assert(x == 0 || x == 1);
+        double log_denom = log(N + alpha + beta);
+        double log_numer = x ? log(s + alpha) : log(N - s + beta);
+        return log_numer - log_denom;
+    }
+    double logp_score() const {
+        double v1 = lbeta(s + alpha, N - s + beta);
```
Comment (alex-lew) [inline cxx/distributions/beta_bernoulli.hh:45]: Our current declaration / implementation of `lbeta` in `util_math.hh` and `util_math.cc` declare the arguments as `int`s. My compiler issued warnings that floating-point numbers were being coerced. I think now that we have hyperparameters that can be non-integer values, we'll want to change the argument types of `lbeta` (I think the implementation should stay the same).

File: cxx/distributions/dirichlet_categorical.hh [hunk 2]
```diff
@@ -3,57 +3,81 @@
...
+        return double(idx);
+    }
+    void transition_hyperparameters() {
+      std::vector<double> logps;
+      std::vector<double> alphas;
+      // C++ doesn't yet allow range for-loops over existing variables.  Sigh.
+      for (double alphat : ALPHA_GRID) {
+          alpha = alphat;
+          double lp = logp_score();
+          if (!std::isnan(lp)) {
```
Comment (alex-lew) [inline cxx/distributions/dirichlet_categorical.hh:75]: Do you have a sense of when nans are showing up?

**Review by alex-lew — APPROVED** [review]
The math looks good! Left a couple minor questions.

File: cxx/distributions/dirichlet_categorical.hh (hunk 2 above)
Comment (ThomasColthurst) [reply cxx/distributions/dirichlet_categorical.hh:75]: None in dirichlet_categorical, at least with the current unit test, but I did see them previously in BetaBernoulli's transition_hyperparamters.

Huh.  When I rerun that now, I no longer see any Nan's.  I wonder if the above lbeta change fixed them.

**Pull request merged**